### DISCUSSION TITLE
How to exclude fields from method return value based on query parameters?

### DISCUSSION THREAD

**Discussion by Cat7373** [prisma/prisma#23947] [Q&A] [answered]
### Question

My model is defined as follows:

```prisma
model User {
  id       Int      @id @default(autoincrement())
  name     String   @unique
  password String
  regionId Int
  role     Role
  addtime  DateTime
  regUid   Int?
  lock     Boolean
}
```

When querying using findMany, the return value is defined as follows:

```typescript
const r = await prisma.user.findMany({ omit: { password: true } })
```

![image](https://github.com/prisma/prisma/assets/9296576/91727dca-6bfa-44b7-b34b-4cab20e2289a)

As you can see, there is no password field.

---

I'm trying to write a general utility method whose return value is deduced as follows:

```typescript
async function testQuery<M, A extends Prisma.Args<M, 'findMany'>>(model: M, queryParams: A) {
  return await (model as any).findMany({ ...queryParams }) as Result<M, A, 'findMany'>
}

const r = await testQuery(usePrisma().user, { omit: { password: true } })
```

![image](https://github.com/prisma/prisma/assets/9296576/97c47913-c03f-43a3-a0b9-de415133a78b)

As you can see, the password field was not successfully excluded.

Please help, how to define the parameters and return values of testQuery so that parameter exclusion can be handled correctly, thank you.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["omitApi"]
}

datasource db {
  provider          = "mysql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}

enum Role {
  ADMIN
  USER
}

model User {
  id       Int      @id @default(autoincrement())
  name     String   @unique
  password String
  regionId Int
  role     Role
  addtime  DateTime
  regUid   Int?
  lock     Boolean
}
```

- OS: MacOS
- Database: MySQL
- Node.js version: v21.7.3

```
Environment variables loaded from .env
prisma                  : 5.13.0
@prisma/client          : 5.13.0
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v21.7.3
Query Engine (Node-API) : libquery-engine b9a39a7ee606c28e3455d0fd60e78c3ba82b1a2b (at node_modules/.pnpm/@prisma+engines@5.13.0/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli b9a39a7ee606c28e3455d0fd60e78c3ba82b1a2b (at node_modules/.pnpm/@prisma+engines@5.13.0/node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.13.0-23.b9a39a7ee606c28e3455d0fd60e78c3ba82b1a2b
Default Engines Hash    : b9a39a7ee606c28e3455d0fd60e78c3ba82b1a2b
Studio                  : 0.500.0
Preview Features        : omitApi
```

**Comment by Cat7373** [comment]
Update, setting the select field is valid:

```typescript
const r = await testQuery(usePrisma().user, { select: { id: true, name: true } })
```

<img width="331" alt="image" src="https://github.com/prisma/prisma/assets/9296576/78f3e3f4-df3f-4369-9f4f-9661cff2c287">

In other words, it may be that the new omit has not been adapted yet.

**Comment by ludralph** [comment]
Hi @Cat7373 👋 

Thank you for raising this question.

We are trying to reproduce this but are running into a type error for `Result` type defined in your code. Can you share the `Result` type definition so we can reproduce this?

**Comment by SevInf** [comment, marked as answer]
First, I can confirm the issue on our side. The problem seems to be TS infering type of `A` to be broader that it actually is. Second, I have a workaround for you until the problem is fixed from our side. You need to add a `const` modifier to `A` in your `testQuery` definition, like this:

```ts
async function testQuery<M, const A extends Prisma.Args<M, 'findMany'>>(model: M, queryParams: A) {
  return (await (model as any).findMany({ ...queryParams })) as Prisma.Result<M, A, 'findMany'>
}
```

**Comment by janpio** [comment]
This discussion has been turned into a bug issue: https://github.com/prisma/prisma/issues/23990
